Implemented changes to psi.lean file.

Strichcoder opened the pull request:
This is about issue #73  [Proving slash action relations for the $\psi$ functions](https://github.com/thefundamentaltheor3m/Sphere-Packing-Lean/issues/73)

Also, the sign typo in the blueprint is fixed, see Zulip thread: [#Sphere packing in 8 dimensions > Typo in Blueprint](https://leanprover.zulipchat.com/#narrow/channel/509682-Sphere-packing-in-8-dimensions/topic/Typo.20in.20Blueprint)

Closes #73

Strichcoder commented:
awaiting-review

seewoo5 commented:
Have you checked my recent comment in #113? I think you don't need to make new PR if that guess is correct.

Strichcoder commented:
yes, I had checked your comment and had tried to do what you said. It did kind of work, but I messed up the merge process.. I would like to just do it cleanly here.

seewoo5 commented on SpherePacking/MagicFunction/b/psi.lean:
```diff
@@ -51,66 +51,414 @@ def ψT' (z : ℂ) : ℂ := if hz : 0 < z.im then ψT ⟨z, hz⟩ else 0
+-- this completes the proof of ψI_eq
+
+-- the following two Lemma statements were wrong in this file at first:
+-- lemma ψS_eq : ψS = 128 * ((H₃_MF + H₄_MF) / (H₂_MF ^ 2) + (H₂_MF + H₃_MF) / H₄_MF ^ 2) := by
+-- lemma ψT_eq : ψT = 128 * ((H₄_MF - H₂_MF) / (H₃_MF ^ 2) - (H₂_MF + H₃_MF) / H₄_MF ^ 2
+--   + (H₂_MF + H₃_MF) / H₄_MF ^ 2) := by
```
Same here

seewoo5 commented on SpherePacking/MagicFunction/b/psi.lean:
```diff
@@ -51,66 +51,408 @@ def ψT' (z : ℂ) : ℂ := if hz : 0 < z.im then ψT ⟨z, hz⟩ else 0
+  rw [hz] at hh
+  exact (lt_self_iff_false 0).mp hh
+
+private lemma slashS (z :ℍ) (F : ℍ → ℂ) : (F ∣[(2 : ℤ)] (S)) (z) =
```
```suggestion
private lemma slashS (z : ℍ) (F : ℍ → ℂ) : (F ∣[(2 : ℤ)] (S)) (z) =
```

seewoo5 commented on SpherePacking/MagicFunction/b/psi.lean:
```diff
@@ -51,66 +51,408 @@ def ψT' (z : ℂ) : ℂ := if hz : 0 < z.im then ψT ⟨z, hz⟩ else 0
+  left
+  rw [pow_coe_nat]
+
+private lemma slashS'' (z : ℍ) (F : ℍ → ℂ): F (S • z) =
```
```suggestion
private lemma slashS'' (z : ℍ) (F : ℍ → ℂ) : F (S • z) =
```

seewoo5 commented on SpherePacking/MagicFunction/b/psi.lean:
```diff
@@ -51,66 +51,408 @@ def ψT' (z : ℂ) : ℂ := if hz : 0 < z.im then ψT ⟨z, hz⟩ else 0
+    rw [zpow_two, pow_two]
+  rw [pow_coe_nat]
+
+private lemma slashST'' (z : ℍ) (F : ℍ → ℂ): F ((S * T) • z) =
```
```suggestion
private lemma slashST'' (z : ℍ) (F : ℍ → ℂ) : F ((S * T) • z) =
```

seewoo5 reviewed (commented):
Some minor lint warnings from [CI](https://github.com/thefundamentaltheor3m/Sphere-Packing-Lean/actions/runs/16916590418/job/47947006520?pr=144)

Strichcoder commented:
Thanks, I implemented the suggestions (and cleaned up the comments a bit more) Then when I wanted to `git push origin branch-psi`, it did not work, so I did `git push origin branch-psi --force`, and it says I changed a whole bunch of files! This was probably bad? Is it possible to revert this? Should I revert it?

seewoo5 commented:
@Strichcoder Yes you can update.

thefundamentaltheor3m commented on SpherePacking/MagicFunction/b/psi.lean:
```diff
@@ -51,71 +51,406 @@ def ψT' (z : ℂ) : ℂ := if hz : 0 < z.im then ψT ⟨z, hz⟩ else 0
+  have hh3 : (H₃_MF : ℍ → ℂ) = H₃ := by exact rfl
+  have hh4 : (H₄_MF : ℍ → ℂ) = H₄ := by exact rfl
+  rw [hh2, hh3, hh4, H₂_T_action, H₃_T_action, H₄_T_action]
+  field_simp -- if I replace this with the result of simp? the proof breaks
+  rw [← mul_div_assoc', ← mul_div_assoc', ←  mul_add, add_comm (H₄ z) (H₃ z),
+    add_comm (H₃ z) (H₂ z) ]
```
```suggestion
  field_simp [← mul_div_assoc', ← mul_add, add_comm (H₄ z) (H₃ z), add_comm (H₃ z) (H₂ z)]
```

thefundamentaltheor3m commented on SpherePacking/MagicFunction/b/psi.lean:
```diff
@@ -51,71 +51,406 @@ def ψT' (z : ℂ) : ℂ := if hz : 0 < z.im then ψT ⟨z, hz⟩ else 0
+  have hh3 : (H₃_MF : ℍ → ℂ) = H₃ := by exact rfl
+  have hh4 : (H₄_MF : ℍ → ℂ) = H₄ := by exact rfl
+  rw [hh2, hh3, hh4, H₂_T_action, H₃_T_action, H₄_T_action]
+  field_simp -- if I replace this by the result of simp?, the proof no longer works.
+  rw [← mul_div_assoc', ← mul_div_assoc', ← mul_add, add_comm (H₄ z) (H₃ z),
+    add_comm  (- (H₂ z)) (H₄ z), Mathlib.Tactic.RingNF.add_neg]
```
You can merge these two lines either by incorporating things in the second line into the first, as I have done below, or by doing `field_simp` followed by `ring_nf`. Both of these would be acceptable.

A small remark: lemmas like `Mathlib.Tactic.RingNF.add_neg` should not be referenced explicitly---not here or anywhere else. These are lemmas that are specifically meant to be used by the `ring_nf` tactic. In fact, if you go to its proof, you will find that it is really just `sub_eq_add_neg`.

```suggestion
  field_simp [← mul_div_assoc', ← mul_add, add_comm (H₄ z) (H₃ z), add_comm  (- (H₂ z)) (H₄ z),
    sub_eq_add_neg]
```

thefundamentaltheor3m commented on SpherePacking/MagicFunction/b/psi.lean:
```diff
@@ -51,71 +51,406 @@ def ψT' (z : ℂ) : ℂ := if hz : 0 < z.im then ψT ⟨z, hz⟩ else 0
+      mul_comm h3 d, mul_assoc d h3 (d * h3), ← div_div, ← mul_div_assoc',
+      div_self, mul_one, mul_comm d h3, ← mul_assoc, ←  div_div, ← pow_two]
+    exact d_nonzero
+  rw [resolve]
```
Same problem with `resolve`

thefundamentaltheor3m commented on SpherePacking/MagicFunction/b/psi.lean:
```diff
@@ -51,71 +51,406 @@ def ψT' (z : ℂ) : ℂ := if hz : 0 < z.im then ψT ⟨z, hz⟩ else 0
+    rw [← right_distrib, pow_two, mul_comm h4 d, mul_assoc d h4 (d * h4), ← div_div,
+       ← mul_div_assoc', div_self, mul_one, mul_comm d h4, ← mul_assoc,←  div_div, ← pow_two ]
+    exact d_nonzero
+  rw [resolve]
```
Again

thefundamentaltheor3m commented on SpherePacking/MagicFunction/b/psi.lean:
```diff
@@ -51,71 +51,406 @@ def ψT' (z : ℂ) : ℂ := if hz : 0 < z.im then ψT ⟨z, hz⟩ else 0
+      mul_assoc d h2 (d * h2), ← div_div, ← mul_div_assoc', div_self]
+    · rw [mul_one, mul_comm d h2, ← mul_assoc, ← div_div, ← pow_two]
+    · exact d_nonzero
+  rw [resolve]
```
Again, need to fix `resolve`

thefundamentaltheor3m commented on blueprint/src/subsections/construct-a-b.tex:
```diff
@@ -339,7 +339,7 @@
     \psi_I(z) &= 128 \frac{H_3(z) + H_4(z)}{H_2(z)^2} + 128 \frac{H_4(z) - H_2(z)}{H_3(z)^2} \label{eqn:psiI-explicit} \\
     \psi_T(z) &= 128 \frac{H_3(z) + H_4(z)}{H_2(z)^2} + 128 \frac{H_2(z) + H_3(z)}{H_4(z)^2} \label{eqn:psiT-explicit} \\
-    \psi_S(z) &= 128 \frac{H_2(z) + H_3(z)}{H_4(z)^2} - 128 \frac{H_2(z) - H_4(z)}{H_3(z)^2} \label{eqn:psiS-explicit}
+    \psi_S(z) &= -128 \frac{H_2(z) + H_3(z)}{H_4(z)^2} - 128 \frac{H_2(z) - H_4(z)}{H_3(z)^2} \label{eqn:psiS-explicit}
```
Good catch! I noticed something like this at some point when working on my thesis, but forgot to update the blueprint. Thank you for doing this!

seewoo5 replied:
Just in case, I also added zulip link in the comment above.

Strichcoder commented:
awaiting-review

thefundamentaltheor3m commented:
Looks great! I agree that it might be possible to golf proofs, but this is not as important as making sure that there are no non-terminal `simp`s, etc. Thank you very much @Strichcoder and @seewoo5!
